### PR TITLE
Document how to change base type of an enum

### PR DESCRIPTION
**Pull request by Blacksmoke16**
* ~Add `TypeNode.base_type` for `Enum` types~
* ~Add `Enum.base_type` that uses the macro method~
* Mention the ability to change the base type in the `Enum` API docs

### CODE REVIEW

**Comment by Blacksmoke16** [comment]
Umm, not sure why specs failed.  They passed locally for me.  Some caching issue?

File: src/compiler/crystal/macros.cr [hunk 1]
```diff
@@ -1860,6 +1860,28 @@ module Crystal::Macros
...
+    # enum IOMode : UInt16
+    #   Read
+    #   Write
+    #   Async
+    # end
+    #
+    # {{Color.base_type}}  # => Int32
+    # {{IOMode.base_type}} # => UInt16
+    # ```
+    def base_type : TypeNode
```
Comment (asterite) [inline src/compiler/crystal/macros.cr:1882]: I would call it `enum_base_type` because otherwise someone might confuse it with `superclass`.

**Comment by asterite** [comment]
std specs also run with the current compiler, and `base_type` is not defined there. You can implement it by doing `typeof(values.first)`

**Comment by Blacksmoke16** [comment]
@asterite But how are all the other macro methods useable? `{% if @type.has_attribute?("Flags") %}` is a `TypeNode` method and that doesn't cause things to break.

**Comment by straight-shoota** [comment]
Can you show a use case for this? `typeof(value)` already works to determine the base type. So this is essentially just adding a dedicated method for this? Where is this needed?

**Comment by Blacksmoke16** [comment]
@straight-shoota It originally came from some chat in Gitter.  https://gitter.im/crystal-lang/crystal?at=5f7afbbc6e0eb8446967c41e

I'd maybe ask @HertzDevil what he was doing it for.  But IMO, there should be a better API for getting that type than doing a `typeof` of one of the values given it's known at compile time anyway.

**Comment by straight-shoota** [comment]
Unless there's a common use case I'd say `type(value)` is totally fine. No need for adding an extra part to the API. `typeof` is evaluated at compile time, so there's no runtime impact.

**Comment by Blacksmoke16** [comment]
Will leave it up to @HertzDevil then.  If he doesn't have any specific use cases then I'm fine with closing, or at least revert the new method and keep the doc updates.

**Comment by Blacksmoke16** [comment]
I kinda forgot about this, but I just reverted the methods and just made this document how to change the base type of an enum.

File: src/enum.cr [hunk 2]
```diff
@@ -86,6 +86,22 @@
...
+# enum Color : UInt8
+#   Red
+#   Green
+#   Blue
+# end
+#
+# Color::Red.value # => UInt8
+# ```
+#
+# NOTE: Only `Int` based types are allowed.
```
Comment (straight-shoota) [inline src/enum.cr:104]: The type must be a primitive, not any `Int`. `BigInt` is not possible for example.
```suggestion
# NOTE: Only `Int::Primitive ` types are allowed.
```

**Review by straight-shoota — APPROVED** [review]

**Review by asterite — APPROVED** [review]

**Review by straight-shoota — APPROVED** [review]

**Comment by HertzDevil** [comment]
There is actually a use case for getting the base type of an enum at compilation time.

`Atomic(T)#max` calls either `Ops.atomicrmw(:max, ...)` or `Ops.atomicrmw(:umax, ...)` depending on whether `T` is a signed type; this decision must be made at compilation time, because `atomicrmw` expects a constant for the operation kind. Currently `#max` checks this by doing `T < Int::Signed`, but since enum types are not subtypes of primitive integers, all atomic enums will use the unsigned variant instead:

```crystal
enum Foo : Int32
  X = Int32::MAX
  Y = Int32::MIN
end

foo = Atomic.new(Foo::X)
foo.min(Foo::Y)
foo.get.value # => 2147483647
foo.max(Foo::Y)
foo.get.value # => -2147483648
```

**Pull request merged**